src/index.js

```diff
@@ -0,0 +1,85 @@
+const crypto = require('crypto');
+const assert = require('assert');
+
+function string2Boolean(str) {
+  return (typeof str === 'boolean' ? `${str}` : str);
+}
+
+function boolean2String(bool) {
+  return (['true', 'false'].includes(bool) ? bool === 'true' : bool);
+}
+
+module.exports = {
+  /**
+   * Encrypt Array of strings or string
+   * @param  {Array|string} data - data to encrypt
+   * @param  {string} key        - key used for encryption
+   * @return {Array|object}      - Array of objects with encrypted data or one object
+   */
+  encrypt: (data, key) => {
+    // if already encrypted
+    if ([null, undefined].includes(data) || [null, undefined].includes(key)) {
+      return data;
+    }
+    
+    const isArray = Array.isArray(data);
+    const queue = isArray ? [...data] : [data];
+    const result = [];
+    
+    for (let item of queue) {
+      if (item.content && item.vector && item.tag) {
+        result.push(item);
+        continue; // eslint-disable-line no-continue
+      }
+
+      const itemValue = string2Boolean(item);
+      const vector = crypto.randomBytes(12);
+      const cipher = crypto.createCipheriv('aes-256-gcm', key, vector);
+      const encryptedItem = `${cipher.update(itemValue, 'utf8', 'hex')}${cipher.final('hex')}`;
+      const tag = cipher.getAuthTag();
+      result.push({
+        content: encryptedItem,
+        vector: vector.toString('hex'),
+        tag: tag.toString('hex')
+      });
+    }
+    
+    return isArray ? result : result[0];
+  },
+  /**
+   * Decrypt Array of objects or one object of the following structure:
+   * {
+   *   content: {string}, - encrypted content
+   *   vector: {string},  - encryption vector
+   *   tag: {string}      - encryption auth tag
+   * }
+   * @param  {Array|string} hash - encrypted data
+   * @param  {string} key        - key used for decryption
+   * @return {Array|string}      - Array of decrypted strings or decrypted string
+   */
+  decrypt: (data, key) => {
+    if ([null, undefined].includes(data) || [null, undefined].includes(key)) {
+      return data;
+    }
+    
+    const isArray = Array.isArray(data);
+    const queue = isArray ? [...data] : [data];
+    const result = [];
+    
+    for (let item of queue) {
+      // if was not encrypted
+      if (!item.content || !item.vector || !item.tag) {
+        result.push(item);
+        continue; // eslint-disable-line no-continue
+      }
+
+      const decipher = crypto.createDecipheriv('aes-256-gcm', key, Buffer.from(item.vector, 'hex'));
```

Comment:
@williamluang  can we further simplify bi hiding the vector from user (so he/she does not need to think about it) and reuse on module level between encrypt and decrypt functions?

Reply:
What do you mean by `reuse on module level` ?

Reply:
to extract into a varaiable with enough scope to be used in both functions, would that work ?

Reply:
Hm.. I don't think so because from the decrypt function, the vector is from the data it receives as argument

Reply:
I think it can since `const vector = crypto.randomBytes(12);` you can just move it to be in scope of both functions and reuse.

Reply:
Oh ok I see ! I didn't understand what you wanted to extract.
Sorry, yes it would definitely works